fish-rust/src/common.rs

```diff
@@ -1,13 +1,1707 @@
+        const NO_PRINTABLES = 1 << 0;
+        /// Do not try to use 'simplified' quoted escapes, and do not use empty quotes as the empty
+        /// string.
+        const NO_QUOTED = 1 << 1;
+        /// Do not escape tildes.
+        const NO_TILDE = 1 << 2;
+        /// Replace non-printable control characters with Unicode symbols.
+        const SYMBOLIC = 1 << 3;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum UnescapeStringStyle {
+    Script(UnescapeFlags),
+    Url,
+    Var,
+}
+
+impl Default for UnescapeStringStyle {
+    fn default() -> Self {
+        Self::Script(UnescapeFlags::default())
+    }
+}
+
+bitflags! {
+    /// Flags for unescape_string functions.
+    #[derive(Default)]
+    pub struct UnescapeFlags: u32 {
+        /// default behavior
+        const DEFAULT = 0;
+        /// escape special fish syntax characters like the semicolon
+        const SPECIAL = 1 << 0;
+        /// allow incomplete escape sequences
+        const INCOMPLETE = 1 << 1;
+        /// don't handle backslash escapes
+        const NO_BACKSLASHES = 1 << 2;
+    }
+}
+
+/// Replace special characters with backslash escape sequences. Newline is replaced with `\n`, etc.
+pub fn escape_string(s: &wstr, style: EscapeStringStyle) -> WString {
+    match style {
+        EscapeStringStyle::Script(flags) => escape_string_script(s, flags),
+        EscapeStringStyle::Url => escape_string_url(s),
+        EscapeStringStyle::Var => escape_string_var(s),
+        EscapeStringStyle::Regex => escape_string_pcre2(s),
+    }
+}
+
+/// Escape a string in a fashion suitable for using in fish script.
+#[widestrs]
+fn escape_string_script(input: &wstr, flags: EscapeFlags) -> WString {
+    let escape_printables = !flags.contains(EscapeFlags::NO_PRINTABLES);
+    let no_quoted = flags.contains(EscapeFlags::NO_QUOTED);
+    let no_tilde = flags.contains(EscapeFlags::NO_TILDE);
+    let no_qmark = feature_test(FeatureFlag::qmark_noglob);
+    let symbolic = flags.contains(EscapeFlags::SYMBOLIC) && MB_CUR_MAX() > 1;
+
+    assert!(
+        !symbolic || !escape_printables,
+        "symbolic implies escape-no-printables"
+    );
+
+    let mut need_escape = false;
+    let mut need_complex_escape = false;
+
+    if !no_quoted && input.is_empty() {
+        return "''"L.to_owned();
+    }
+
+    let mut out = WString::new();
+
+    for c in input.chars() {
+        if c >= ENCODE_DIRECT_BASE && c < ENCODE_DIRECT_END {
+            let val = wchar_decode_byte(c);
+            out += "\\X";
+
+            let nibble1 = val / 16;
+            let nibble2 = val % 16;
+
+            out.push(char::from_digit(nibble1.into(), 16).unwrap());
+            out.push(char::from_digit(nibble2.into(), 16).unwrap());
+            need_escape = true;
+            need_complex_escape = true;
+            continue;
+        }
+        match c {
+            '\t' => {
+                if symbolic {
+                    out.push('␉');
```

Comment:
```suggestion
                    out.push('\t');
```

same below, having these raw characters in source code isn't very nice

Reply:
they are not the same. A literal tab would render as whitespace (of variable width).
This is from the Unicode Control Pictures, it's a symbol for tab that conveniently fits into one display column https://www.fileformat.info/info/unicode/char/2409/index.htm
This is used for displaying control characters to the user in the completion pager

Reply:
Ooh, tricky, thanks for clearing that up.